xarray/core/indexing.py

```diff
@@ -177,18 +206,36 @@ def convert_label_indexer(index, label, index_name='', method=None,
             if np.any(indexer < 0):
                 raise KeyError('not all values found in index %r'
                                % index_name)
-    return indexer
+    return indexer, new_index
 
 
 def remap_label_indexers(data_obj, indexers, method=None, tolerance=None):
     """Given an xarray data object and label based indexers, return a mapping
-    of equivalent location based indexers.
+    of equivalent location based indexers. Also return a mapping of pandas'
+    single index objects returned from multi-index objects.
     """
     if method is not None and not isinstance(method, str):
         raise TypeError('``method`` must be a string')
-    return dict((dim, convert_label_indexer(data_obj[dim].to_index(), label,
-                                            dim, method, tolerance))
-                for dim, label in iteritems(indexers))
+
+    pos_indexers, new_indexes = {}, {}
+    for dim, label in iteritems(indexers):
+        index = data_obj[dim].to_index()
+
+        if isinstance(index, pd.MultiIndex):
+            # set default names for multi-index unnamed levels so that
+            # we can safely rename dimension / coordinate later
```

Comment:
Compared to `as_compatible_data`, moving to `to_index()` would create a copy of the multi-index (almost free I guess) each time it is called but it would also allow tracking level name changes on the original `MultiIndex`:

```python
idx = pd.MultiIndex.from_product([['a', 'b'], [1, 2], [-1, -2]])
>>> y = xr.DataArray(np.random.rand(2 * 2 * 2), [('x', idx)])
>>> y.x.to_index()
MultiIndex(levels=[['a', 'b'], [1, 2], [-2, -1]],
           labels=[[0, 0, 0, 0, 1, 1, 1, 1], [0, 0, 1, 1, 0, 0, 1, 1], [1, 0, 1, 0, 1, 0, 1, 0]],
           names=['x_level_0', 'x_level_1', 'x_level_2'])
>>> idx.names = ('one', 'two', 'three')
>>> y.x.to_index()
MultiIndex(levels=[['a', 'b'], [1, 2], [-2, -1]],
           labels=[[0, 0, 0, 0, 1, 1, 1, 1], [0, 0, 1, 1, 0, 0, 1, 1], [1, 0, 1, 0, 1, 0, 1, 0]],
           names=['one', 'two', 'three'])
```

and we have also direct access to the dimension/coordinate name so that we can set default level names that are unique (as shown above).